### PR TITLE
Cache Serialize API

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Adding constructs to support reservations over cached objects. This is usefully to mitigate duplicate work over long running, frequent tasks. Specifically we added:
- PhaseWaitingForCache: to indicate that a task is currently waiting for the cache to be populated
- API extensions to datacatalog client for handling cache reservations

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
RFC - https://github.com/flyteorg/flyte/pull/1461

## Follow-up issue
_NA_

### CODE REVIEW

**Review by kumare3 — COMMENTED** [review]
are there changes to be made for array tasks?

**Comment by EngHabu** [comment]
That's a great question, @kumare3. I vote for no and call this out in the docs of this feature... that the way it works for array jobs is on the entire input of the array job as opposed to individual elements... the reason is that the array job code is complex enough that I don't think we should introduce additional complexity with this... until we move to array nodes...

**Review by EngHabu — APPROVED** [review]

File: go/tasks/pluginmachinery/catalog/client.go [hunk 1]
```diff
@@ -73,10 +75,55 @@ func NewCatalogEntry(outputs io.OutputReader, status Status) Entry {
 	return Entry{outputs: outputs, status: status}
 }
 
+// Indicates the Reservation in Catalog
+type ReservationEntry struct {
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/catalog/client.go:79]: Maybe add a few unit tests for this, admittedly trivial, file?

While we are at it, can we add docs on public types/funcs/methods?

File: go/tasks/pluginmachinery/catalog/client.go [hunk 2]
```diff
@@ -73,10 +75,55 @@ func NewCatalogEntry(outputs io.OutputReader, status Status) Entry {
 	return Entry{outputs: outputs, status: status}
 }
 
+// Indicates the Reservation in Catalog
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/catalog/client.go:78]: nit: golang-style ... it seems

```suggestion
// ReservationEntry indicates the Reservation in Catalog
```

File: go/tasks/pluginmachinery/catalog/client.go (hunk 1 above)
Comment (hamersaw) [reply go/tasks/pluginmachinery/catalog/client.go:79]: done

File: go/tasks/pluginmachinery/catalog/client.go (hunk 2 above)
Comment (hamersaw) [reply go/tasks/pluginmachinery/catalog/client.go]: done

**Review by EngHabu — APPROVED** [review]
